core/include/drivers/i2c.h

```diff
@@ -0,0 +1,346 @@
+ *
+ * @__name: I2C device driver name
+ * @__match_table: match table associated to the driver
+ * @__i2c_probe: I2C probe function with the following prototype:
+ *	TEE_Result (*probe)(struct i2c_dev *i2c_dev, const void *fdt,
+ *			    int node, const void *compat_data);
+ */
+#define DEFINE_I2C_DEV_DRIVER(__name, __match_table, __i2c_probe) \
+	static TEE_Result __name ## _probe_i2c_dev(const void *fdt, int node, \
+						   const void *compat_data) \
+	{ \
+		struct i2c_dev *i2c_dev = NULL; \
+		TEE_Result res = TEE_ERROR_GENERIC; \
+		res = i2c_dt_get_dev(fdt, node, &i2c_dev); \
+		if (res) \
+			return res; \
+		return __i2c_probe(i2c_dev, fdt, node, compat_data); \
+	} \
+	DEFINE_DT_DRIVER(__name ## _dt_driver) = { \
+		.name = # __name, \
+		.type = DT_DRIVER_I2C, \
+		.match_table = __match_table, \
+		.probe = __name ## _probe_i2c_dev, \
+	}
+
+#define I2C_SMBUS_MAX_BUF_SIZE	32
+
+enum i2c_smbus_dir {
+	I2C_SMBUS_READ,
+	I2C_SMBUS_WRITE,
+};
+
+enum i2c_smbus_protocol {
+	I2C_SMBUS_PROTO_BYTE,
+	/* Like block but does not insert "count" in sent data, useful for
+	 * EEPROM read for instance which is not SMBus but requires such
+	 * sequence.
+	 */
+	I2C_SMBUS_PROTO_BLOCK_RAW,
+};
+
+struct i2c_ctrl;
+
+/**
+ * struct i2c_dev - I2C device
+ *
+ * @ctrl: I2C controller associated to the device
+ * @addr: device address on the I2C bus
+ * @priv: Private data associated to the device
+ */
+struct i2c_dev {
+	struct i2c_ctrl *ctrl;
+	uint8_t addr;
+	void *priv;
+};
+
+/**
+ * struct i2c_ctrl_ops - Operations provided by I2C controller drivers
+ *
+ * @read: I2C read operation
+ * @write: I2C write operation
+ * @smbus: SMBus protocol operation
+ */
+struct i2c_ctrl_ops {
+	TEE_Result (*read)(struct i2c_dev *i2c_dev, uint8_t *buf, int len);
+	TEE_Result (*write)(struct i2c_dev *i2c_dev, const uint8_t *buf,
+			    int len);
+	TEE_Result (*smbus)(struct i2c_dev *i2c_dev, uint8_t dir, uint8_t proto,
+			    uint8_t cmd_code, uint8_t *buf, unsigned int len);
+};
+
+/**
+ * struct i2c_ctrl - I2C controller
+ *
+ * @ops: Operations associated to the I2C controller
+ * @priv: Private data associated to the controller
+ */
+struct i2c_ctrl {
+	const struct i2c_ctrl_ops *ops;
+	void *priv;
+};
+
+#ifdef CFG_DRIVERS_I2C
+/**
+ * i2c_create_dev - Create and i2c_dev struct from device-tree
+ *
+ * @i2c_ctrl: Controller to be used with this device
+ * @fdt: Device-tree to work on
+ * @node: Node to work on in @fdt provided device-tree
+ *
+ * Return an i2c_dev struct filled from device-tree description
+ */
+struct i2c_dev *i2c_create_dev(struct i2c_ctrl *i2c_ctrl, const void *fdt,
+			       int node);
+
+/**
+ * i2c_write() - Execute an I2C write on the I2C bus
+ *
+ * @i2c_dev: I2C device used for writing
+ * @buf: Buffer of data to be written
+ * @len: Length of data to be written
+ *
+ * Return a TEE_Result compliant value
+ */
+static inline TEE_Result i2c_write(struct i2c_dev *i2c_dev, const uint8_t *buf,
+				   int len)
```

Comment:
`size_t`

Reply:
Acked, will change that for all `len` passing